### PR TITLE
Implement optimization bisect

### PR DESCRIPTION
**Pull request by tugsbayasgalan**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#49031 Implement optimization bisect**

Differential Revision: [D25691790](https://our.internmc.facebook.com/intern/diff/D25691790)

### CODE REVIEW

File: torch/csrc/jit/jit_opt_bisect.h [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+// `TorchScript` offers a simple bisection utility function
+// that can be configured through environment variable `PYTORCH_OPT_LIMIT`.
+// The purpose is to limit how many optimization you can make per pass.
+// This is useful for debugging performance pass related pass.
+
+// Bisecting is enabled on a per file basis (hence per pass). For example, in
+// `constant_propagation.cpp`, `PYTORCH_OPT_LIMIT` should be
+// set to `constant_propagation.cpp=<opt_limt>` or, simply, to
+// `constant_propagation.cpp=<opt_limit>` where <opt_limit> is the number of
```
Comment (ZolotukhinM) [inline torch/csrc/jit/jit_opt_bisect.h:15]: Nit: did you mean `constant_propagation=<opt_limit>` here (without `.cpp`)?

File: torch/csrc/jit/jit_opt_bisect.h [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+    int l,
+    const std::string& in_str);
+
+TORCH_API bool is_bisect_enabled(const char* pass_name);
+
+TORCH_API std::ostream& operator<<(std::ostream& out, int64_t level);
+
+#define JIT_BISECT(...) (is_bisect_enabled(__FILE__));
+
+#define JIT_BISECT_LOG(MSG, ...)                           \
```
Comment (ZolotukhinM) [inline torch/csrc/jit/jit_opt_bisect.h:54]: What's the difference between `JIT_BISECT_LOG` and `GRAPH_DUMP` and friends? If possible, we should avoid duplicating functionality.

File: torch/csrc/jit/passes/constant_propagation.cpp [hunk 3]
```diff
@@ -336,6 +338,13 @@ struct ConstantPropagator {
   }
 
   void ConstantPropagation(Node* n) {
+    auto allowed = JIT_BISECT();
+    if (!allowed) {
+      JIT_BISECT_LOG(
```
Comment (ZolotukhinM) [inline torch/csrc/jit/passes/constant_propagation.cpp:343]: This will probably spam the output after the optimization has reached its limit (i.e. we will keep trying to run the optimization no matter what, but with opt limit after some point each run would simply print this message and return). However, it might useful to detect the **last** invocation of the optimization - would that be possible to do somehow?

File: torch/csrc/jit/jit_opt_bisect.cpp [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+    return n;
+  } catch (...) {
+    return -1;
+  }
+}
+
+static std::unordered_map<std::string, int64_t> parseJITBisectOption(
+    const char* option) {
+  std::stringstream in_ss;
+  in_ss << "function:";
```
Comment (ZolotukhinM) [inline torch/csrc/jit/jit_opt_bisect.cpp:32]: What is this "function:"? I didn't see it mentioned in the comment.

File: torch/csrc/jit/jit_opt_bisect.cpp [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+
+  std::unordered_map<std::string, int64_t> passes_to_opt_limits;
+  std::string line;
+  while (std::getline(in_ss, line, ':')) {
+    if (line.size() == 0) {
+      continue;
+    }
+    auto index_at = line.find_last_of('=');
+    auto pass_name = line.substr(0, index_at);
+    pass_name = c10::detail::ExcludeFileExtension(pass_name);
```
Comment (ZolotukhinM) [inline torch/csrc/jit/jit_opt_bisect.cpp:45]: Since we're adding a `ExcludeFileExtension` function, it would make sense to also update this place:
https://github.com/pytorch/pytorch/blob/fb755ad33ef84844e4f4c2d285b71737316814f8/torch/csrc/jit/jit_log.cpp#L45-L48

It also might be a good idea to extract it to a separate PR, but up to you.

File: torch/csrc/jit/jit_opt_bisect.h [hunk 6]
```diff
@@ -0,0 +1,60 @@
+#pragma once
+#include <torch/csrc/WindowsTorchApiMacro.h>
+#include <memory>
+#include <string>
+#include <unordered_map>
+
+// `TorchScript` offers a simple bisection utility function
+// that can be configured through environment variable `PYTORCH_OPT_LIMIT`.
```
Comment (ZolotukhinM) [inline torch/csrc/jit/jit_opt_bisect.h:8]: Nit: since it only affect JIT passes, it might makes sense to include `JIT` into the env var name.

File: torch/csrc/jit/jit_opt_bisect.h [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+
+TORCH_API std::string jit_bisect_prefix(
+    const std::string& msg,
+    const char* fn,
+    int l,
+    const std::string& in_str);
+
+TORCH_API bool is_bisect_enabled(const char* pass_name);
+
+TORCH_API std::ostream& operator<<(std::ostream& out, int64_t level);
```
Comment (ZolotukhinM) [inline torch/csrc/jit/jit_opt_bisect.h:50]: Why do we need to redefine this operator for a standard type? Also, it doesn't seem to be implemented anywhere.

File: torch/csrc/jit/jit_opt_bisect.cpp [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+#include <torch/csrc/jit/frontend/error_report.h>
+#include <torch/csrc/jit/ir/ir.h>
+#include <torch/csrc/jit/jit_opt_bisect.h>
+#include <torch/csrc/jit/serialization/python_print.h>
+
+namespace torch {
+namespace jit {
+
+static int parseOptLimit(const std::string& opt_limit) {
+  // TODO bad code
```
Comment (ZolotukhinM) [inline torch/csrc/jit/jit_opt_bisect.cpp:20]: 😃

File: torch/csrc/jit/jit_opt_bisect.cpp [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+#include <torch/csrc/jit/api/function_impl.h>
+#include <torch/csrc/jit/frontend/error_report.h>
+#include <torch/csrc/jit/ir/ir.h>
+#include <torch/csrc/jit/jit_opt_bisect.h>
+#include <torch/csrc/jit/serialization/python_print.h>
+
+namespace torch {
+namespace jit {
+
+static int parseOptLimit(const std::string& opt_limit) {
```
Comment (ZolotukhinM) [inline torch/csrc/jit/jit_opt_bisect.cpp:19]: Nit: since everywhere else we're using `int64_t` it might make sense to return it here too.

File: torch/csrc/jit/jit_opt_bisect.cpp [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+#include <vector>
+
+#include <ATen/core/function.h>
+#include <c10/util/Exception.h>
+#include <c10/util/StringUtil.h>
+#include <torch/csrc/jit/api/function_impl.h>
+#include <torch/csrc/jit/frontend/error_report.h>
+#include <torch/csrc/jit/ir/ir.h>
+#include <torch/csrc/jit/jit_opt_bisect.h>
+#include <torch/csrc/jit/serialization/python_print.h>
```
Comment (ZolotukhinM) [inline torch/csrc/jit/jit_opt_bisect.cpp:14]: What do we need this include for?

File: torch/csrc/jit/jit_opt_bisect.cpp [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+
+  static const std::unordered_map<std::string, int64_t> passes_to_opt_limits =
+      parseJITBisectOption(opt_limit);
+  std::string pass{pass_name};
+  pass = c10::detail::StripBasename(pass);
+  pass = c10::detail::ExcludeFileExtension(pass);
+
+  auto opt_limit_it = passes_to_opt_limits.find(pass);
+  if (opt_limit_it == passes_to_opt_limits.end()) {
+    return false;
```
Comment (ZolotukhinM) [inline torch/csrc/jit/jit_opt_bisect.cpp:68]: Shouldn't we return true (don't limit the optimization) if it was not listed in `PYTORCH_OPT_LIMIT`?

File: torch/csrc/jit/jit_opt_bisect.cpp [hunk 12]
```diff
@@ -0,0 +1,121 @@
...
+  }
+
+  return passes_to_opt_limits;
+}
+
+bool is_bisect_enabled(const char* pass_name) {
+  static const char* opt_limit = std::getenv("PYTORCH_OPT_LIMIT");
+  // if nothing is provided, let's allow everything
+  if (!opt_limit) {
+    return true;
```
Comment (ZolotukhinM) [inline torch/csrc/jit/jit_opt_bisect.cpp:57]: Nit: I find the name of the function a bit confusing: it says "bisect is enabled", but when the `PYTORCH_OPT_LIMIT ` flag is not set, then we return `true`.

File: torch/csrc/jit/jit_opt_bisect.h [hunk 13]
```diff
@@ -0,0 +1,60 @@
+#pragma once
+#include <torch/csrc/WindowsTorchApiMacro.h>
+#include <memory>
+#include <string>
+#include <unordered_map>
+
+// `TorchScript` offers a simple bisection utility function
```
Comment (ZolotukhinM) [inline torch/csrc/jit/jit_opt_bisect.h:7]: We probably also need to add a description to `OVERVIEW.md`: https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/OVERVIEW.md#jit-logging

File: torch/csrc/jit/passes/dead_code_elimination.cpp [hunk 14]
```diff
@@ -259,6 +260,14 @@ class DeadCodeEliminator {
 
   // Delete all unmarked nodes.
   void sweep(Block* block, bool recurse) {
+    auto allowed = JIT_BISECT();
+    if (!allowed) {
+      JIT_BISECT_LOG(
+        std::string("Stop deleting because of opt limitation"),
```
Comment (ZolotukhinM) [inline torch/csrc/jit/passes/dead_code_elimination.cpp:266]: Nit: generally it might be a good idea to "standardize" such messages across passes.

**Review by ZolotukhinM — COMMENTED** [review]
Very nice! I'm super excited to see it's landed and start using it :) I left some comments inline, mostly minor. Once we resolve them, it'll be good to go!

Also, cc @Krovatkin who implemented `jit_log` and might provide some valuable feedback.

File: torch/csrc/jit/jit_opt_bisect.h (hunk 2 above)
Comment (tugsbayasgalan) [reply torch/csrc/jit/jit_opt_bisect.h:54]: Yeah I think we can use GRAPH_DUMP but I think user has to manually pass in the bisect related information to the log in that case.

File: torch/csrc/jit/jit_opt_bisect.cpp (hunk 5 above)
Comment (tugsbayasgalan) [reply torch/csrc/jit/jit_opt_bisect.cpp:45]: sounds good! will do

File: torch/csrc/jit/jit_opt_bisect.cpp (hunk 4 above)
Comment (tugsbayasgalan) [reply torch/csrc/jit/jit_opt_bisect.cpp:32]: oh sorry this was from jit_log.cpp and i forgot to delete

File: torch/csrc/jit/jit_opt_bisect.cpp (hunk 10 above)
Comment (tugsbayasgalan) [reply torch/csrc/jit/jit_opt_bisect.cpp:14]: good catch! will remove them

File: torch/csrc/jit/jit_opt_bisect.cpp (hunk 11 above)
Comment (tugsbayasgalan) [reply torch/csrc/jit/jit_opt_bisect.cpp:68]: Ah yep that is true! will fix

File: torch/csrc/jit/passes/constant_propagation.cpp [hunk 15]
```diff
@@ -336,6 +338,13 @@ struct ConstantPropagator {
   }
 
   void ConstantPropagation(Node* n) {
+    auto allowed = JIT_BISECT();
```
Comment (ZolotukhinM) [inline torch/csrc/jit/passes/constant_propagation.cpp:341]: Nit: a random thought, I've just realized that while this API will probably mostly be used for the bisection, the macro itself doesn't do any bisection: it simply enables/disables the optimization. Maybe we should pick a different name that better reflects what it does?

**Comment by tugsbayasgalan** [comment]
I will submit a seperate PR to use ExcludeFileExtension on jit_log since this PR is not related to jit_log

File: torch/csrc/jit/jit_opt_limit.cpp [hunk 16]
```diff
@@ -0,0 +1,83 @@
...
+#include <c10/util/Exception.h>
+#include <c10/util/StringUtil.h>
+#include <torch/csrc/jit/api/function_impl.h>
+#include <torch/csrc/jit/jit_opt_limit.h>
+
+namespace torch {
+namespace jit {
+
+static int parseOptLimit(const std::string& opt_limit) {
+  // TODO bad code
```
Comment (ZolotukhinM) [inline torch/csrc/jit/jit_opt_limit.cpp:18]: Please address this todo :) Or expand on what exactly needs to be done.

File: torch/csrc/jit/jit_opt_limit.h [hunk 17]
```diff
@@ -0,0 +1,46 @@
+#pragma once
+#include <torch/csrc/WindowsTorchApiMacro.h>
+#include <memory>
+#include <string>
+#include <unordered_map>
+
+// `TorchScript` offers a simple optimization limit checker
+// that can be configured through environment variable `PYTORCH_OPT_LIMIT`.
```
Comment (ZolotukhinM) [inline torch/csrc/jit/jit_opt_limit.h:8]: Nit: `PYTORCH_JIT_OPT_LIMIT`

File: torch/csrc/jit/jit_opt_limit.h [hunk 18]
```diff
@@ -0,0 +1,46 @@
...
+// The purpose is to limit how many optimization you can make per pass.
+// This is useful for debugging performance passes.
```
Comment (ZolotukhinM) [inline torch/csrc/jit/jit_opt_limit.h:10]: Nit: not only performance, any passes :)

File: torch/csrc/jit/passes/constant_propagation.cpp [hunk 19]
```diff
@@ -12,6 +12,7 @@
 #include <torch/csrc/jit/runtime/operator.h>
 #include <torch/csrc/jit/runtime/vararg_functions.h>
 #include <torch/csrc/utils/memory.h>
+#include <string>
```
Comment (ZolotukhinM) [inline torch/csrc/jit/passes/constant_propagation.cpp:15]: Nit: Unnecessary change.

File: torch/csrc/jit/jit_opt_limit.h [hunk 20]
```diff
@@ -0,0 +1,46 @@
...
+// if (!allowed) {
+//     GRAPH_DUMP(...); //supplied from jit_log
+//     return;
+// }
+
+namespace torch {
+namespace jit {
+
+struct Node;
+struct Graph;
```
Comment (ZolotukhinM) [inline torch/csrc/jit/jit_opt_limit.h:37]: Nit: These forward decls are not used.

File: torch/csrc/jit/jit_opt_limit.h [hunk 21]
```diff
@@ -0,0 +1,46 @@
+#pragma once
+#include <torch/csrc/WindowsTorchApiMacro.h>
+#include <memory>
```
Comment (ZolotukhinM) [inline torch/csrc/jit/jit_opt_limit.h:3]: Nit: unused?

**Review by ZolotukhinM — APPROVED** [review]
Awesome work, thanks! I only have some nitpicks :)

**Comment by ZolotukhinM** [comment]
Oh, please don't forget to update the commit message before landing too!

**Review by ZolotukhinM — APPROVED** [review]
👍
